### PR TITLE
[CPU] attn supports f16

### PR DESCRIPTION
**Pull request by xczhai**
### Details:
 - *rebase from https://github.com/openvinotoolkit/openvino/pull/22939*
 - *enable avx512 fp16 for attention*
 - *enable amx fp16 for attention*
 - *update PagedAttentionExtension lightly. can specify the correct type to pa second output precision*

### Tickets:
 - *128183*

### CODE REVIEW

File: src/inference/src/system_conf.cpp [hunk 1]
```diff
@@ -97,7 +101,8 @@ bool with_cpu_x86_avx512_core_amx_fp16() {
 }
 
 bool with_cpu_x86_avx512_core_amx() {
-    return with_cpu_x86_avx512_core_amx_int8() || with_cpu_x86_avx512_core_amx_bf16();
+    return with_cpu_x86_avx512_core_amx_int8() || with_cpu_x86_avx512_core_amx_bf16() ||
```
Comment (luo-cheng2021) [inline src/inference/src/system_conf.cpp:104]: We'd better not to change this function which is consistent with onednn.

File: src/inference/src/system_conf.cpp [hunk 2]
```diff
@@ -80,6 +80,10 @@ bool with_cpu_x86_bfloat16() {
     return get_cpu_info().has(Xbyak::util::Cpu::tAVX512_BF16);
 }
 
+bool with_cpu_x86_avx512_core_bf16() {
```
Comment (luo-cheng2021) [inline src/inference/src/system_conf.cpp:83]: It should be deleted since there is no code reference?

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp [hunk 3]
```diff
@@ -84,18 +105,30 @@ static constexpr size_t vec_len_f32_neon = vec_len_neon / sizeof(float);
...
+        _mm512_mask_storeu_ps(addr, mask_addr, v);
     }
-    inline __m512 mm512_uni_loadu_tail_ps(const ov::float16* a, size_t count) {
-        auto mask = (1 << count) - 1;
-        auto f16_vec = _mm256_maskz_loadu_epi16(mask, a);
-        return _mm512_cvtph_ps(f16_vec);
+
+    inline void mm512_uni_storeu_tail_ps(ov::bfloat16 *addr, __m512 v, size_t count) {
+        __mmask16 mask_addr = (1 << count) - 1;
+        // mm512_uni_mask_storeu_ps(addr, mask_addr, v); // TODO can remove
```
[the remark this answers is not in the document]
Comment (xczhai) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp]: > No use, please remove

removed

File: src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp [hunk 4]
```diff
@@ -38,32 +38,54 @@ BrgemmKernel::BrgemmKernel(size_t M,
     // blocking M
     M_blk = matmulOptimalM;
     M_tail = M % M_blk;
-    brgVnniFactor = 4 / inType.size();
 
-    if (inType != ov::element::bf16 && inType != ov::element::f32)
-        THROW_ERROR("brgemm kernel only supports bf16, f32");
+    if (inType != ov::element::bf16 && inType != ov::element::f32 && inType != ov::element::f16)
```
[the remark this answers is not in the document]
Comment (xczhai) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp]: > Using `!one_of(inType, ...)` should be more graceful.

update it.

File: src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp [hunk 5]
```diff
@@ -238,8 +280,9 @@ void BrgemmKernel::init_brgemm_copy_b(
     bool transpose,
     size_t copy_B_wei_stride) {
     brgemm_matmul_conf_t brgCopyKernelConf;
-    brgCopyKernelConf.src_dt = dt_in0;
-    brgCopyKernelConf.wei_dt = dt_in1;
+    // TODO: AMX_FP16
```
[the remark this answers is not in the document]
Comment (xczhai) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp]: > Remove.

removed

File: src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp [hunk 6]
```diff
@@ -341,17 +389,18 @@ void BrgemmKernel::executeGemm(bool is_M_tail, void* a, void* b, void* c, void*
             size_t mIdx = is_M_tail ? 1 : 0;
             auto& brgemmCtx = brgCtxs[getBrgIdx(mIdx, k, n)];
             if (brgemmCtx.K != 0 && brgemmCtx.N != 0 && brgemmCtx.M != 0) {
-                auto local_a_ptr = k > 0 ? ptr_a_tail : ptr_A;
-                auto B_stride = (k * count_K + n * count_N * brgVnniFactor) * inType.size();
+                // TODO: AMX_FP16
```
[the remark this answers is not in the document]
Comment (xczhai) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp]: > Remove

removed

File: src/plugins/intel_cpu/src/nodes/scaled_attn.cpp [hunk 7]
```diff
@@ -1065,6 +1065,17 @@ void ScaledDotProductAttention::createPrimitive() {
         if (rtPrecision == ov::element::bf16) {
 #ifdef OPENVINO_ARCH_X86_64
             executor = std::make_shared<AttentionExecutor<KT_ONEDNN, ov::bfloat16>>(context);
+#endif
+        } else if (rtPrecision == ov::element::f16) {
+#ifdef OPENVINO_ARCH_X86_64
+            // TODO: AMX_FP16
```
[the remark this answers is not in the document]
Comment (xczhai) [reply src/plugins/intel_cpu/src/nodes/scaled_attn.cpp]: > Remove

removed

File: src/inference/src/system_conf.cpp (hunk 1 above)
Comment (xczhai) [reply src/inference/src/system_conf.cpp]: > We'd better not to change this function which is consistent with onednn.

recover it.

File: src/inference/src/system_conf.cpp (hunk 2 above)
Comment (xczhai) [reply src/inference/src/system_conf.cpp]: > It should be deleted since there is no code reference?

removed.

**Comment by xczhai** [comment]
@luo-cheng2021 Have updated the code according comments. Could you help review the codes, especially page attention part? Thanks.

File: src/plugins/intel_cpu/src/graph.cpp [hunk 8]
```diff
@@ -193,9 +193,12 @@ void Graph::Replicate(const std::shared_ptr<const ov::Model> &model,
         auto parentNode = op2node[unusedOutput.get_node_shared_ptr()];
         const auto port = unusedOutput.get_index();
         const auto nodeName = std::string("stub_") + std::to_string(unusedOutput.get_index()) + "_" + parentNode->getName();
+        // WA: avoid PagedAttention's second output reorder.
```
Comment (luo-cheng2021) [inline src/plugins/intel_cpu/src/graph.cpp:196]: We'd better not to hardcode here and should find the place where the output precision is changed to f16.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 9]
```diff
@@ -851,11 +856,19 @@ struct MHAHelper {
...
+                if (dnnl::impl::cpu::x64::mayiuse(dnnl::impl::cpu::x64::amx_bf16) &&
+                    precision_of<KVCACHE_TYPE>::value == ov::element::bf16) {
+                    _fastpath_valid_prec = ov::element::bf16;
+                } else if (dnnl::impl::cpu::x64::mayiuse(dnnl::impl::cpu::x64::amx_fp16) &&
+                           precision_of<KVCACHE_TYPE>::value == ov::element::f16) {
+                    _fastpath_valid_prec = ov::element::f16;
+                }
+            }
+            if (one_of(_fastpath_valid_prec, ov::element::bf16, ov::element::f16) && !_gemv) {
+                // aligned to cache line (64bytes=16*sizeof(float)) to avoid false sharing
```
Comment (luo-cheng2021) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:869]: Remove the unrelated comment

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp [hunk 10]
```diff
@@ -1136,6 +1161,15 @@ struct MHAHelper {
                         }
                     }
                     _gemv->tile_release();
+                } else if (_fastpath_valid_prec == ov::element::f16) {
```
Comment (luo-cheng2021) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp:1164]: May we merge this branch to the above since the code is similar and no performance benefit?

Comment (xczhai) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp]: > May we merge this branch to the above since the code is similar and no performance benefit?

merge the similar codes.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp (hunk 9 above)
Comment (xczhai) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/executor_pa.cpp]: > Remove the unrelated comment

removed

**Comment by yuxu42** [comment]
Hi @dmitry-gorokhov Could you please take a review? Thanks!

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp [hunk 11]
```diff
@@ -84,18 +105,29 @@ static constexpr size_t vec_len_f32_neon = vec_len_neon / sizeof(float);
...
+        __m512i xpi32 = _mm512_castps_si512(v);
+        __m512i nan = _mm512_set1_epi32(0xffff);
+        auto mask = _mm512_cmp_ps_mask(v, v, _CMP_ORD_Q);
+        __m512i ones = _mm512_set1_epi32(0x1);
+        __m512i vec_bias = _mm512_set1_epi32(0x7fff);
+        auto x = _mm512_and_si512(_mm512_srli_epi32(xpi32, 16), ones); // LSB = x[16]
+        x = _mm512_add_epi32(x, vec_bias);                             // rounding_bias = 0x7fff + LSB
+        x = _mm512_srli_epi32(_mm512_add_epi32(x, xpi32), 16);         // x = (x + rounding_bias) >> 16;
+        x = _mm512_mask_blend_epi32(mask, nan, x);                     // Check NaN before converting back to bf16
+        _mm512_mask_cvtepi32_storeu_epi16(addr, mask_addr, x);
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp:124]: I do not really understand the math in this function. Could you please provide the details?

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp [hunk 12]
```diff
@@ -155,21 +204,30 @@ static constexpr size_t vec_len_f32_neon = vec_len_neon / sizeof(float);
...
-    }
     inline void mm256_uni_storeu_ps(ov::float16* a,  __m256 v) {
         __m128i vec_f16 = _mm256_cvtps_ph(v, 0);
         _mm_storeu_si128(reinterpret_cast<__m128i *>(a), vec_f16);
     }
 
+    // store __m256 to addr
+    inline void mm256_uni_storeu_tail_ps(float *addr, __m256 v, size_t count) {
+        const auto mask = get_mask(count);
+        return _mm256_maskstore_ps(addr, mask, v);
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp:215]: This intrinsic is available in avx512 only. You cannot use it for avx2 code-path

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp [hunk 13]
```diff
@@ -155,21 +204,30 @@ static constexpr size_t vec_len_f32_neon = vec_len_neon / sizeof(float);
...
+    }
+
+    inline void mm256_uni_storeu_tail_ps(ov::bfloat16 *addr, __m256 v, size_t count) {
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp:218]: Sohuld't we disable bf16 template specialization of Attn kernel for avx2 at all? It will allow to avoid implimentation of such functions.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp [hunk 14]
```diff
@@ -249,6 +307,11 @@ static constexpr size_t vec_len_f32_neon = vec_len_neon / sizeof(float);
         float32x4_t vec_f32 = vcvtq_f32_u32(vmovl_u16(vec_bf16));
         return vec_f32;
     }
+    inline float32x4_t __vld1q_f32(const ov::float16* a) {
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp:310]: Why do we need changes in ARM code-path?

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp [hunk 15]
```diff
@@ -672,10 +673,12 @@ inline void multiply_scalar(float* a, ov::bfloat16* a_dst, const float val, cons
...
         v_a = _mm512_maskz_loadu_ps(mask, a + i);
         v_a = _mm512_mul_ps(v_a, v_scale);
-        mm512_uni_mask_storeu_ps(a_dst + i, mask, v_a);
+        mm512_uni_storeu_tail_ps(a_dst + i, v_a, size - i);
+
+        i += (size - i);
     }
 #else
-    for (size_t i = 0; i < size; i++) {
+    for (; i < size; i++) {
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp:681]: Why this change was done? Seems like vector code should handle all the elements.

File: src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp [hunk 16]
```diff
@@ -255,17 +297,23 @@ void BrgemmKernel::init_brgemm_copy_b(
...
+        brgCopyKernelConf.isa = dt_in0 == dnnl_data_type_t::dnnl_f16 ? avx512_core_amx_fp16 : avx512_core_amx;
         brgCopyKernelConf.s8s8_compensation_required = false;
     } else {
-        brgCopyKernelConf.isa = dt_in0 == dnnl_data_type_t::dnnl_bf16 ? avx512_core_bf16 : avx512_core_vnni;
+        if (inType == ov::element::f16) {
+            brgCopyKernelConf.isa = mayiuse(avx512_core_fp16) ? avx512_core_fp16 : avx2_vnni_2;
+        } else {
+            brgCopyKernelConf.isa = dt_in0 == dnnl_data_type_t::dnnl_bf16 ? avx512_core_bf16 : avx512_core_vnni;
+        }
+        brgCopyKernelConf.s8s8_compensation_required = dt_in0 == dnnl_data_type_t::dnnl_s8;
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp:316]: Why do we need this? Based on the code above brgemm kernel supports only floating point precisions.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/common/concat_multiple_query_sdp.cpp [hunk 17]
```diff
@@ -72,7 +74,8 @@ class ConcatMultiQuerySDPTest : public testing::WithParamInterface<ConcatMultiQu
             result << ")_";
         }
         result << "Prc=" << qkvType << "_";
-        result << "HasShapeOf=" << hasShapeof << "_";
+        result << "ForceKVU8" << forceKVU8 << "_";
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/common/concat_multiple_query_sdp.cpp:77]: How about using `ElementType kvCacheType;` for better understanding?

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/common/concat_multiple_query_sdp.cpp [hunk 18]
```diff
@@ -290,17 +304,24 @@ class ConcatMultiQuerySDPTest : public testing::WithParamInterface<ConcatMultiQu
 TEST_P(ConcatMultiQuerySDPTest, CompareWithRefs) {
     SKIP_IF_CURRENT_TEST_IS_DISABLED();
     InputShapeAndTransposeOrder inputShapeAndOrders;
+    bool forceKVU8;
     bool hasShapeOf;
     ElementType qkvType;
-    std::tie(qkvType, inputShapeAndOrders, hasShapeOf) = this->GetParam();
-    if (qkvType == ElementType::bf16 && !ov::with_cpu_x86_bfloat16())
+    std::tie(qkvType, inputShapeAndOrders, forceKVU8, hasShapeOf) = this->GetParam();
+    if ((qkvType == ElementType::bf16 && !ov::with_cpu_x86_bfloat16()) ||
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/common/concat_multiple_query_sdp.cpp:311]: Can we instans only valid test params combinations instead of implicit skips inside the test body?

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp (hunk 11 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp:124]: The code is same for the following(plus additional logic to keep nan):
https://github.com/openvinotoolkit/openvino/blob/46a6ccd4ed93d36dd24183b69bd64204c634debe/src/core/include/openvino/core/type/bfloat16.hpp#L85-L87
@usstq please correct me if I'm wrong.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp (hunk 12 above)
Comment (luo-cheng2021) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp:215]: `_mm256_maskstore_ps` will be converted to `vmaskmovps` which should be exist in [AVX](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#ig_expand=5312,3117,5354,5638,4165,2329,4675,4675,2840,7545,7197,7208,7200,7197,4786,907,551,2329,1618,3335,4484,2232,2672,3434,3485,4646,504,713,1887,2689,6951,1878,1619,2520,2499,1947,2125,2073,4918,7084,4895,4918,1640,6596,506,4083,1641,699,2959,2977,2978,4989,4993,4994,5014,5018,6118,4588,536,598,2871,2961,1927,1926,2727,1641,1640,2701,2707,2707,4635,1762,4635,2945,4279,4280,4281,1690,1690,1690,3981,4857,4880,4871,4112,4088,4111,7030,7033,4111,1866,1664,7130,7131,6039,1866,3103,1511,1622,2174,4871,1640,2502,1869,4111,750,92,4030,1866,1664,3107,5749,750,1920,1929,6551,6551,4111,4044,140,607,668,822,3062,4360,4687,5821,6633,2245,1866,7131,2055,1916,1916,2055,2110,2052,4208,2738,4252,4263,4603,6220,460,4263,2227,6408,1753,1664,6875,1774,6176,1774,5822,545,1869,1667,5303,2955,2246,694,4019,6568,6555,5152,750,512,6877,6882,6883,5152,5152,5154,5155,6863,1931,1919,1572,6518,4082,6550,4975,5062,1572,2317,6550,5050,3878,691,7009,4108,5821,3111,6517,6053,6050,4928,5056,4928,6027,4930,7054,7054,7057,4986,7057,6015,4986,4934,4927,5053,5053,4953,4953,4954,4940,6015,6053,6033,2946,4954,4954,6047,4949,4954,4949,4954,2664,6015,4930,6015,4987,5062,5002,5062,6561,6560,6564,2570,2107,2110,1869,1667,1929,2668,2123,1918,1897,1920,1900,1923,1911,4106,1869,1667,574,5788,2632,2958,1927,1926,1900,1919,4030,4016,4108,4108,5643,6248,564,3841,4064,4110,4110,6233,5643,244,5665,5647,828,4263&amxtechs=AMX_BF16,AMX_INT8,AMX_TILE,AMX_FP16,AMX_COMPLEX&techs=SSE_ALL,AVX_ALL,AVX_512&text=_mm256_maskstore_ps).

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp (hunk 14 above)
Comment (xczhai) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp:310]: It is to fix a compilation error. 
`scale_add2_reduce_max` 's template `T` can be `float16` and it calls `__vld1q_f32` in arm codes.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp (hunk 15 above)
Comment (xczhai) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/softmax_kernel.hpp:681]: > Why this change was done? Seems like vector code should handle all the elements.

Actually, no any big change here. Just use a template to unify the `bfloat16` and `float16` code together.

File: src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp (hunk 13 above)
Comment (xczhai) [reply src/plugins/intel_cpu/src/nodes/kernels/scaled_attn/common.hpp]: > Sohuld't we disable bf16 template specialization of Attn kernel for avx2 at all? It will allow to avoid implimentation of such functions.

Yes. You are right. This function is redundant and should not be invoked. Remove

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/common/concat_multiple_query_sdp.cpp (hunk 17 above)
Comment (xczhai) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/common/concat_multiple_query_sdp.cpp]: > How about using `ElementType kvCacheType;` for better understanding?

accept it.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/common/concat_multiple_query_sdp.cpp (hunk 18 above)
Comment (xczhai) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/common/concat_multiple_query_sdp.cpp:312]: > Can we instans only valid test params combinations instead of implicit skips inside the test body?

It is a runtime aware testcase, especially for different ISA.
This test body will not run on unsupported platform. I think it it relates to ISA instead of params.

File: src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp (hunk 16 above)
Comment (xczhai) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/brgemm_kernel.cpp:316]: > Why do we need this? Based on the code above brgemm kernel supports only floating point precisions.

It is always `false`. Just leave it for future s8 supporting. Remove it.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/common/concat_multiple_query_sdp.cpp (hunk 18 above)
Comment (dmitry-gorokhov) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/common/concat_multiple_query_sdp.cpp:312]: We already have generic mechanism that skips all the bf16 tests on non avx512 platforms. https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/tests/functional/shared_tests_instances/skip_tests_config.cpp#L518-L523
So I don't understand why this test has to be implemented in different way

**Comment by dmitry-gorokhov** [comment]
@xczhai I am going to merge ARM SDPA FP16 first https://github.com/openvinotoolkit/openvino/pull/26487. Lets rebase the PR after that.

File: src/plugins/intel_cpu/src/graph.cpp (hunk 8 above)
Comment (xczhai) [reply src/plugins/intel_cpu/src/graph.cpp:196]: > Any updates here. It has to be resolved before the merge.

still address it. The root cause is `ConvertPrecision` transformation. Need to deal with it carefully and avoid affecting gpu.

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/common/concat_multiple_query_sdp.cpp (hunk 18 above)
Comment (xczhai) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/common/concat_multiple_query_sdp.cpp]: > We already have generic mechanism that skips all the bf16 tests on non avx512 platforms. https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_cpu/tests/functional/shared_tests_instances/skip_tests_config.cpp#L518-L523 So I don't understand why this test has to be implemented in different way

I see. remove the skip codes.

File: src/plugins/intel_cpu/src/graph.cpp (hunk 8 above)
Comment (xczhai) [reply src/plugins/intel_cpu/src/graph.cpp]: @dmitry-gorokhov @luo-cheng2021 
remove such wa.
The root cause is that the PagedAttentionExtension op doesn't provide a mechanism to change the 2nd output dtype.
I make some changes.
1. add a `set_out_type` member func to PagedAttentionExtension op.
2. When execute `validate_and_infer_types` in PagedAttentionExtension, it will determine output type. It won't break the GPU path.
3. add a `fuse_type_to_pa` in CPU plugin, which is a extend to `ConvertPrecision`. It is used to specify the correct type for `PagedAttentionExtension` 's 2nd output type.

File: src/core/src/op/paged_attention.cpp [hunk 19]
```diff
@@ -146,13 +146,26 @@ void PagedAttentionExtension::validate_and_infer_types() {
...
+        set_output_type(1, m_output_type[1], {Dimension::dynamic()});
+    }
 }
 
 std::shared_ptr<ov::Node> PagedAttentionExtension::clone_with_new_inputs(const ov::OutputVector& new_args) const {
     return std::make_shared<PagedAttentionExtension>(new_args);
 }
 
+void PagedAttentionExtension::set_out_type(int index, const ov::element::Type& output_type) {
+    m_output_type[index] = output_type;
```
Comment (luo-cheng2021) [inline src/core/src/op/paged_attention.cpp:167]: Better to check if `index` is valid.

**Review by luo-cheng2021 — APPROVED** [review]
LGTM.

File: src/plugins/intel_cpu/src/graph.cpp [hunk 20]
```diff
@@ -193,9 +193,10 @@ void Graph::Replicate(const std::shared_ptr<const ov::Model> &model,
         auto parentNode = op2node[unusedOutput.get_node_shared_ptr()];
         const auto port = unusedOutput.get_index();
         const auto nodeName = std::string("stub_") + std::to_string(unusedOutput.get_index()) + "_" + parentNode->getName();
+        auto nodePrec = parentNode->getOriginalOutputPrecisionAtPort(port);
```
[the remark this answers is not in the document]
Comment (xczhai) [reply src/plugins/intel_cpu/src/graph.cpp]: > Better revert to the original version since there is no useful changes.

accept. recover it.

**Comment by xczhai** [comment]
> @xczhai I am going to merge ARM SDPA FP16 first #26487. Lets rebase the PR after that.

has

File: src/plugins/intel_cpu/src/nodes/scaled_attn.cpp [hunk 21]
```diff
@@ -1071,6 +1071,16 @@ void ScaledDotProductAttention::createPrimitive() {
         if (rtPrecision == ov::element::bf16) {
 #ifdef OPENVINO_ARCH_X86_64
             executor = std::make_shared<AttentionExecutor<KT_ONEDNN, ov::bfloat16>>(context);
+#endif
+        } else if (rtPrecision == ov::element::f16) {
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/scaled_attn.cpp:1075]: Seems like it it makes inaccessible branch on line 1088.
Maybe this is the reason why Arm test is failing

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp [hunk 22]
```diff
@@ -201,6 +210,17 @@ std::vector<ov::Tensor> ConcatSDPTest::run_test(std::shared_ptr<ov::Model> model
...
     SKIP_IF_CURRENT_TEST_IS_DISABLED();
+    ElementType inType;
+    std::vector<InputShape> inputShapes;
+    bool forceKVU8;
+    bool hasShapeOf;
+    std::tie(inType, inputShapes, forceKVU8, hasShapeOf) = this->GetParam();
+
+    if ((inType == ElementType::bf16 && !ov::with_cpu_x86_bfloat16()) ||
+        (inType == ElementType::f16 && !ov::with_cpu_x86_avx512_core_fp16())) {
+        GTEST_SKIP();
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp:221]: Test skip is still there

File: src/plugins/intel_cpu/src/graph.cpp (hunk 8 above)
Comment (dmitry-gorokhov) [reply src/plugins/intel_cpu/src/graph.cpp]: @xczhai Just for my better understanding: could you please descrive the pattern there Reorder is inserted? Like next op after PA expected fp16 on its input?

File: src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp (hunk 22 above)
Comment (xczhai) [reply src/plugins/intel_cpu/tests/functional/custom/subgraph_tests/src/classes/concat_sdp.cpp]: > Test skip is still there

remove

File: src/plugins/intel_cpu/src/nodes/scaled_attn.cpp (hunk 21 above)
Comment (xczhai) [reply src/plugins/intel_cpu/src/nodes/scaled_attn.cpp:1074]: > Seems like it it makes inaccessible branch on line 1088. Maybe this is the reason why Arm test is failing

reorg this code snippet. 
new code looks more clear for different platforms.

Comment (dmitry-gorokhov) [reply src/plugins/intel_cpu/src/nodes/scaled_attn.cpp:1074]: Thanks!

**Review by dmitry-gorokhov — APPROVED** [review]

File: src/plugins/intel_cpu/src/graph.cpp (hunk 8 above)
Comment (xczhai) [reply src/plugins/intel_cpu/src/graph.cpp]: > @xczhai Just for my better understanding: could you please descrive the pattern there Reorder is inserted? Like next op after PA expected fp16 on its input?

Okay.

1. At the very beginning, the PA op spec describes the two outputs type is aligned with input0 type. As a result, PA's outputs type is `f32` when entering CPU plugin.
2. During CPU plugin transformation,  `ConvertPrecision` will convert or fuse the op's type. As a result, PA's two outputs type is `f16`. But remember PA's 2nd output is dangle without any child and `Result` node.
3. In construct graph, all the `dangle` output will be wrapped by `Result` node and the type is aligned with output type. In this case, the specific pattern is `PA's 2nd output` --> `Result(f16)`
4. But in CPU node design, PA's 2nd output is always `f32`. So the pattern is `PA's 2nd output(f32)` --> `Result(f16)`.
5. The following ResolveConflict logic scan this pattern and then insert a `Reorder`. So the pattern becomes `PA's 2nd output(f32)` --> `Reorder` --> `Result(f16)`
